activesupport/lib/active_support/cache/serializer_with_fallback.rb

```diff
@@ -0,0 +1,174 @@
+        end
+
+        def try_compress(dumped, threshold)
+          if dumped.bytesize >= threshold
+            compressed = compress(dumped)
+            compressed unless compressed.bytesize >= dumped.bytesize
+          end
+        end
+
+        def decompress(compressed)
+          Zlib::Inflate.inflate(compressed)
+        end
+
+        module PassthroughWithFallback
+          include SerializerWithFallback
+          extend self
+
+          def dump(entry)
+            entry
+          end
+
+          def dump_compressed(entry, threshold)
+            entry.compressed(threshold)
+          end
+
+          def _load(entry)
+            entry
+          end
+
+          def dumped?(dumped)
+            dumped.is_a?(Cache::Entry)
+          end
+        end
+
+        module Marshal61WithFallback
+          include SerializerWithFallback
+          extend self
+
+          MARSHAL_SIGNATURE = "\x04\x08".b.freeze
+
+          def dump(entry)
+            Marshal.dump(entry)
+          end
+
+          def dump_compressed(entry, threshold)
+            Marshal.dump(entry.compressed(threshold))
+          end
+
+          def _load(dumped)
+            Marshal.load(dumped)
+          end
+
+          def dumped?(dumped)
+            dumped.start_with?(MARSHAL_SIGNATURE)
+          end
+        end
+
+        module Marshal70WithFallback
+          include SerializerWithFallback
+          extend self
+
+          MARK_UNCOMPRESSED = "\x00".b.freeze
+          MARK_COMPRESSED   = "\x01".b.freeze
+
+          def dump(entry)
+            MARK_UNCOMPRESSED + Marshal.dump(entry.pack)
+          end
+
+          def dump_compressed(entry, threshold)
+            dumped = Marshal.dump(entry.pack)
+            if compressed = try_compress(dumped, threshold)
+              MARK_COMPRESSED + compressed
+            else
+              MARK_UNCOMPRESSED + dumped
+            end
+          end
+
+          def _load(marked)
+            dumped = marked.byteslice(1..-1)
+            dumped = decompress(dumped) if marked.start_with?(MARK_COMPRESSED)
+            Cache::Entry.unpack(Marshal.load(dumped))
+          end
+
+          def dumped?(dumped)
+            dumped.start_with?(MARK_UNCOMPRESSED, MARK_COMPRESSED)
+          end
+        end
+
+        module MessagePackWithFallback
+          include SerializerWithFallback
+          extend self
+
+          def dump(entry)
+            ActiveSupport::MessagePack::CacheSerializer.dump(entry.pack)
+          end
+
+          def _load(dumped)
+            packed = ActiveSupport::MessagePack::CacheSerializer.load(dumped)
+            Cache::Entry.unpack(packed) if packed
+          end
+
+          def dumped?(dumped)
+            available? && ActiveSupport::MessagePack.signature?(dumped)
+          end
+
+          private
+            def available?
+              return @available if defined?(@available)
+              require "active_support/message_pack"
```

Comment:
This effectively makes `msgpack` 1.7+ a dependency of Active Support. If you don't have it installed, you will see this line throughout your logs:

```
ActiveSupport::MessagePack requires the msgpack gem, version 1.7.0 or later. Please add it to your Gemfile: `gem "msgpack", ">= 1.7.0"`
```

`msgpack` is already a dependency of `bootsnap` which is included in the default Rails gemfile. So many people will already have it installed, but maybe not the correct version.

@jonathanhefner what do you think about adding `gem "msgpack", ">= 1.7.0"` to the Active Support gemspec, rather than making every user do it?

Reply:
I can make a PR but figured it was easier to ask first.

Reply:
If that's not an option, then  it would be good if you could use Rails caching without seeing that log line, if you don't want to use the message pack serializer.

Reply:
Ahh, right!  Originally it was just a `LoadError`, but then we added a warning (ref: https://github.com/rails/rails/pull/47770#discussion_r1169054263).  Thank you for catching that!

I agree that this line should not display a warning.  To that end, I've submitted #48125 and added you as a co-author.

Reply:
Thanks!